motoko/passwords/src/passwords/main.mo

```diff
@@ -0,0 +1,121 @@
+import Array "mo:base/Array";
+import Principal "mo:base/Principal";
+import Option "mo:base/Option";
+
+actor {
+
+  // Types
+
+  type Credentials = {
+    username : Text;
+    password : Text;
+  };
+
+  type Account = {
+    id : Nat;
+    username : Text;
+    hash : Text;
+    salt : Text;
+  };
+
+  type Session = {
+    principal : Principal;
+    account_id : Nat;
+  };
+
+
+  // Data Structures
+
+  // Arrays a sub-optimal choice here but they're easy to understand
+  var accounts : [Account] = [];
+  var sessions : [Session] = [];
+  var next_account_id : Nat = 0;
+
+  // Public Functions
+
+  public func signup (credentials : Credentials) {
+    // Generate a random salt and hash it together with the password
+    let _salt = random_salt();
+    let _hash = sha256(credentials.password # _salt);
+
+    // Create a new account with the username and hashed password
+    let account = {
+      id = next_account_id;
+      username = credentials.username;
+      hash = _hash;
+      salt = _salt;
+    };
+
+    // Save the account and auto-increment the next account ID
+    accounts := Array.append<Account>(accounts, [account]);
+    next_account_id += 1;
+  };
+
+  public shared ({ caller }) func login (credentials : Credentials) {
+    // Find the account associated with the username
+    let account = Option.unwrap(Array.find(accounts, findByUsername(credentials.username)));
+
+    // If the hashed passwords match then create a session
+    if (sha256(credentials.password # account.salt) == account.hash) {
+      // The session associates this caller's ID with the account
+      let session : Session = {
+        principal = caller;
+        account_id = account.id;
+      };
+      sessions := Array.append(sessions, [session]);
+    };
+  };
+
+  public shared ({ caller }) func logout () {
+    // Remove the session for this caller's ID
+    sessions := Array.filter(sessions, omitCaller(caller));
```

Comment:
cycle cost will be O(n), with HashMap closer to O(1), and also I think it's fine to keep using arrays. (may be worthwhile to educate on the cost-downsides in a comment or something)